Composer/packages/extensions/visual-designer/src/editors/ObiEditor.tsx

```diff
@@ -144,108 +86,90 @@ export const ObiEditor: FC<ObiEditorProps> = ({
-            .then(copiedActions => {
-              const lgTemplatesToBeCreated: { name: string; body: string }[] = [];
-              walkLgResourcesInActionList(copiedActions, (designerId, actionData, fieldName, lgStr) => {
-                if (!lgStr) return '';
-
-                const lgName = new LgMetaData(fieldName, designerId).toString();
-                const refString = new LgTemplateRef(lgName).toString();
-
-                lgTemplatesToBeCreated.push({ name: lgName, body: lgStr });
-                actionData[fieldName] = refString;
-                return refString;
-              });
-              return onCreateDialog(copiedActions).then(dialogName => ({ dialogName, lgTemplatesToBeCreated }));
-            })
-            .then(async ({ dialogName: newDialog, lgTemplatesToBeCreated }) => {
-              // defense modal cancellation
-              if (!newDialog) return;
-
-              // create lg templates for actions in new dialog
-              for (const { name, body } of lgTemplatesToBeCreated) {
-                await updateLgTemplate(newDialog, name, body);
-              }
-
-              // delete old actions (they are already moved to new dialog)
-
-              // HACK: https://github.com/microsoft/BotFramework-Composer/issues/2247
-              const postponedDeleteLgTemplates = templates => setTimeout(() => deleteLgTemplates(templates), 501);
-              const deleteResult = deleteNodes(data, e.actionIds, nodes =>
-                deleteActions(nodes, postponedDeleteLgTemplates, deleteLuIntents)
-              );
-
-              // insert a BeginDialog action points to newly created dialog
-              const indexes = e.actionIds[0].match(/^(.+)\[(\d+)\]$/);
-              if (indexes === null || indexes.length !== 3) return;
-
-              const [, arrayPath, actionIndexStr] = indexes;
-              const startIndex = parseInt(actionIndexStr);
-              const placeholderAction = dialogFactory.create(SDKKinds.BeginDialog, { dialog: newDialog });
-              const insertResult = insertAction(deleteResult, arrayPath, startIndex, placeholderAction);
-              onChange(insertResult);
-            });
+          // Create target dialog
+          const newDialogId = await createDialog();
+          if (!newDialogId) return;
+          const newDialogData = readDialog(newDialogId);
+
+          // Using copy->paste->delete pattern is safer than using cut->paste
+          const actionsToBeMoved = await copySelectedActions(path, data, e.actionIds);
+          const newDialogWithActions = await insertActions(
+            newDialogId,
+            newDialogData,
+            `${'triggers'}[0].${'actions'}`,
+            0,
+            actionsToBeMoved
+          );
+          if (actionsContainLuIntent(actionsToBeMoved)) {
+            // auto assign recognizer type to lu
+            newDialogWithActions.recognizer = `${newDialogId}.lu`;
```

Comment:
recognizer logic leak. Consider adding a `updateRecognizer()` method in `dialogEditApi`